# HG changeset patch
# Parent 7902a1cd526a6a03e9e914e4cfda6f70eb2a27f7

diff --git a/osaf/libs/common/cpsv/include/cpd_cb.h b/osaf/libs/common/cpsv/include/cpd_cb.h
--- a/osaf/libs/common/cpsv/include/cpd_cb.h
+++ b/osaf/libs/common/cpsv/include/cpd_cb.h
@@ -180,7 +180,6 @@ typedef struct cpd_cb_tag {
 	bool is_quiesced_set;
 	MDS_DEST loc_cpnd_dest;
 	MDS_DEST rem_cpnd_dest;
-	/*MDS_DEST             cpnd_dests[NCS_MAX_SLOTS]; */
 
 	NCS_PATRICIA_TREE cpnd_tree;
 
diff --git a/osaf/libs/common/mqsv/include/mqsv_common.h b/osaf/libs/common/mqsv/include/mqsv_common.h
--- a/osaf/libs/common/mqsv/include/mqsv_common.h
+++ b/osaf/libs/common/mqsv/include/mqsv_common.h
@@ -86,7 +86,7 @@
 
 uint32_t machineEndianness(void);
 uint32_t mqsv_listenerq_msg_send(SaMsgQueueHandleT listenerhandle);
-NCS_PHY_SLOT_ID mqsv_get_phy_slot_id(MDS_DEST dest);
+uint32_t mqsv_get_phy_slot_id(MDS_DEST dest);
 typedef unsigned int MSG_FRMT_VER;
 
 #define m_MQSV_CONVERT_SATIME_TEN_MILLI_SEC(t)      (t)/(10000000)	/* 10^7 */
diff --git a/osaf/libs/common/mqsv/mqsv_common.c b/osaf/libs/common/mqsv/mqsv_common.c
--- a/osaf/libs/common/mqsv/mqsv_common.c
+++ b/osaf/libs/common/mqsv/mqsv_common.c
@@ -117,12 +117,13 @@ uint32_t mqsv_listenerq_msg_send(SaMsgQu
 
 *************************************************************************************/
 
-NCS_PHY_SLOT_ID mqsv_get_phy_slot_id(MDS_DEST dest)
+uint32_t mqsv_get_phy_slot_id(MDS_DEST dest)
 {
 	NCS_PHY_SLOT_ID phy_slot;
 	NCS_SUB_SLOT_ID sub_slot;
 
 	m_NCS_GET_PHYINFO_FROM_NODE_ID(m_NCS_NODE_ID_FROM_MDS_DEST(dest), NULL, &phy_slot, &sub_slot);
 
-	return ((sub_slot * NCS_SUB_SLOT_MAX) + phy_slot);
+	osafassert(sub_slot < NCS_SUB_SLOT_MAX);
+	return ((uint32_t) sub_slot) + ((uint32_t) phy_slot) * NCS_SUB_SLOT_MAX;
 }
diff --git a/osaf/libs/core/include/ncsgl_defs.h b/osaf/libs/core/include/ncsgl_defs.h
--- a/osaf/libs/core/include/ncsgl_defs.h
+++ b/osaf/libs/core/include/ncsgl_defs.h
@@ -73,13 +73,13 @@ extern "C" {
 
 /*************************************************
  * Maximum Slots (Including sub slots) supported
- * 16 slots x 8 subslots
+ * 256 slots x 16 subslots
  *************************************************/
-#define NCS_SLOT_MAX 16
+#define NCS_SLOT_MAX 256
 
 #define NCS_SUB_SLOT_MAX 16
 
-#define NCS_MAX_SLOTS ((NCS_SLOT_MAX *  NCS_SUB_SLOT_MAX) + NCS_SLOT_MAX)
+#define NCS_MAX_SLOTS (NCS_SLOT_MAX *  NCS_SUB_SLOT_MAX)
 
 	typedef uint64_t MDS_DEST;
 	typedef uint32_t NCS_NODE_ID;
diff --git a/osaf/libs/core/mds/include/mds_dt.h b/osaf/libs/core/mds/include/mds_dt.h
--- a/osaf/libs/core/mds/include/mds_dt.h
+++ b/osaf/libs/core/mds/include/mds_dt.h
@@ -247,13 +247,20 @@ bool mdtm_mailbox_mbx_cleanup(NCSCONTEXT
 
 #if (MDS_USE_SUBSLOT_ID == 0)
 #define MDS_TIPC_NODE_ID_MIN     0x01001001
-#define MDS_TIPC_NODE_ID_MAX     0x010010ff
-#define MDS_NCS_NODE_ID_MIN      (MDS_NCS_CHASSIS_ID|0x0000010f)
-#define MDS_NCS_NODE_ID_MAX      (MDS_NCS_CHASSIS_ID|0x0000ff0f)
-#define m_MDS_GET_NCS_NODE_ID_FROM_TIPC_NODE_ID(node) \
-        (NODE_ID)( MDS_NCS_CHASSIS_ID | (((node)&0xff)<<8) | (0xf))
-#define m_MDS_GET_TIPC_NODE_ID_FROM_NCS_NODE_ID(node) \
-        (NODE_ID)( MDS_TIPC_COMMON_ID | (((node)&0xff00)>>8) )
+#define MDS_TIPC_NODE_ID_MAX     0x01001fff
+static inline NODE_ID m_MDS_GET_NCS_NODE_ID_FROM_TIPC_NODE_ID(NODE_ID node) {
+        return MDS_NCS_CHASSIS_ID | ((node & 0xff) << 8) | (0xf - ((node & 0xf00) >> 8));
+}
+static inline NODE_ID m_MDS_GET_TIPC_NODE_ID_FROM_NCS_NODE_ID(NODE_ID node) {
+        return MDS_TIPC_COMMON_ID | ((node & 0xff00) >> 8) | ((0xf - (node & 0xf)) << 8);
+}
+static inline uint32_t m_MDS_CHECK_TIPC_NODE_ID_RANGE(NODE_ID node) {
+	return node < MDS_TIPC_NODE_ID_MIN || node > MDS_TIPC_NODE_ID_MAX ?
+		NCSCC_RC_FAILURE : NCSCC_RC_SUCCESS;
+}
+static inline uint32_t m_MDS_CHECK_NCS_NODE_ID_RANGE(NODE_ID node) {
+    return m_MDS_CHECK_TIPC_NODE_ID_RANGE(m_MDS_GET_TIPC_NODE_ID_FROM_NCS_NODE_ID(node));
+}
 #else
 #define MDS_TIPC_NODE_ID_MIN     0x01001001
 #define MDS_TIPC_NODE_ID_MAX     0x0100110f
@@ -263,10 +270,10 @@ bool mdtm_mailbox_mbx_cleanup(NCSCONTEXT
         (NODE_ID)( MDS_NCS_CHASSIS_ID | ((node)&0xf) | (((node)&0xff0)<<4))
 #define m_MDS_GET_TIPC_NODE_ID_FROM_NCS_NODE_ID(node) \
         (NODE_ID)( MDS_TIPC_COMMON_ID | (((node)&0xff00)>>4) | ((node)&0xf) )
-#endif
 
 #define m_MDS_CHECK_TIPC_NODE_ID_RANGE(node) (((((node)<MDS_TIPC_NODE_ID_MIN)||((node)>MDS_TIPC_NODE_ID_MAX))?NCSCC_RC_FAILURE:NCSCC_RC_SUCCESS))
 #define m_MDS_CHECK_NCS_NODE_ID_RANGE(node) (((((node)<MDS_NCS_NODE_ID_MIN)||((node)>MDS_NCS_NODE_ID_MAX))?NCSCC_RC_FAILURE:NCSCC_RC_SUCCESS))
+#endif
 
 /* ******************************************** */
 /* ******************************************** */
diff --git a/osaf/services/infrastructure/nid/scripts/configure_tipc.in b/osaf/services/infrastructure/nid/scripts/configure_tipc.in
--- a/osaf/services/infrastructure/nid/scripts/configure_tipc.in
+++ b/osaf/services/infrastructure/nid/scripts/configure_tipc.in
@@ -61,9 +61,9 @@ if ! test -f "$SLOT_ID_FILE"; then
    exit 1
 fi
 SLOT_ID=`cat "$SLOT_ID_FILE"`
-if [ "$SLOT_ID" -gt "255" ] || [ "$SLOT_ID" -lt "1" ]  
+if [ "$SLOT_ID" -gt "255" ] || [ "$SLOT_ID" -lt "0" ]
     then 
-      echo "SLOT ID Should be in the range of 1 to 255"
+      echo "SLOT ID Should be in the range of 0 to 255"
       echo "Quitting......"
       exit 1
 fi
@@ -77,6 +77,11 @@ if [ "$SUBSLOT_ID" -gt "15" ] || [ "$SUB
     echo "Quitting......"
     exit 1
 fi
+if [ "$SUBSLOT_ID" -eq "15" ] && [ "$SLOT_ID" -eq "0" ]; then
+    echo "The combination SUBSLOT ID=15 and SLOT ID=0 is invalid"
+    echo "Quitting......"
+    exit 1
+fi
 
 if [ "$#" -lt "1" ] ||  [ "$#" -gt "4" ]
    then
@@ -109,7 +114,9 @@ if [ "$USE_SUBSLOT_ID" = "YES" ]; then
     TIPC_NODEID=$(($SLOT_ID << $SHIFT4))
     TIPC_NODEID=$(($TIPC_NODEID + $SUBSLOT_ID))
 else    
-    TIPC_NODEID=$SLOT_ID    
+    TIPC_NODEID=$((15 - SUBSLOT_ID))
+    TIPC_NODEID=$((TIPC_NODEID << 8))
+    TIPC_NODEID=$((TIPC_NODEID + SLOT_ID))
 fi
 
 function tipc_configure ()
diff --git a/osaf/services/saf/cpsv/cpd/cpd_db.c b/osaf/services/saf/cpsv/cpd/cpd_db.c
--- a/osaf/services/saf/cpsv/cpd/cpd_db.c
+++ b/osaf/services/saf/cpsv/cpd/cpd_db.c
@@ -1213,7 +1213,8 @@ uint32_t cpd_get_slot_sub_id_from_mds_de
 
 	m_NCS_GET_PHYINFO_FROM_NODE_ID(m_NCS_NODE_ID_FROM_MDS_DEST(dest), NULL, &phy_slot, &sub_slot);
 
-	return ((sub_slot * NCS_SUB_SLOT_MAX) + (phy_slot));
+	osafassert(sub_slot < NCS_SUB_SLOT_MAX);
+	return ((uint32_t) sub_slot) + ((uint32_t) phy_slot) * NCS_SUB_SLOT_MAX;
 }
 
 /********************************************************************************
@@ -1232,8 +1233,8 @@ uint32_t cpd_get_slot_sub_slot_id_from_n
 
 	m_NCS_GET_PHYINFO_FROM_NODE_ID(i_node_id, NULL, &phy_slot, &sub_slot);
 
-	return ((sub_slot * NCS_SUB_SLOT_MAX) + (phy_slot));
-
+	osafassert(sub_slot < NCS_SUB_SLOT_MAX);
+	return ((uint32_t) sub_slot) + ((uint32_t) phy_slot) * NCS_SUB_SLOT_MAX;
 }
 
 /*******************************************************************************************
diff --git a/osaf/services/saf/cpsv/cpnd/cpnd_db.c b/osaf/services/saf/cpsv/cpnd/cpnd_db.c
--- a/osaf/services/saf/cpsv/cpnd/cpnd_db.c
+++ b/osaf/services/saf/cpsv/cpnd/cpnd_db.c
@@ -733,7 +733,8 @@ uint32_t cpnd_get_slot_sub_slot_id_from_
 
 	m_NCS_GET_PHYINFO_FROM_NODE_ID(m_NCS_NODE_ID_FROM_MDS_DEST(dest), NULL, &phy_slot, &sub_slot);
 
-	return ((sub_slot * NCS_SUB_SLOT_MAX) + (phy_slot));
+	osafassert(sub_slot < NCS_SUB_SLOT_MAX);
+	return ((uint32_t) sub_slot) + ((uint32_t) phy_slot) * NCS_SUB_SLOT_MAX;
 }
 
 /***********************************************************************************
@@ -749,8 +750,8 @@ uint32_t cpnd_get_slot_sub_slot_id_from_
 
 	m_NCS_GET_PHYINFO_FROM_NODE_ID(i_node_id, NULL, &phy_slot, &sub_slot);
 
-	return ((sub_slot * NCS_SUB_SLOT_MAX) + (phy_slot));
-
+	osafassert(sub_slot < NCS_SUB_SLOT_MAX);
+	return ((uint32_t) sub_slot) + ((uint32_t) phy_slot) * NCS_SUB_SLOT_MAX;
 }
 
 /******************************************************************************************
diff --git a/osaf/services/saf/immsv/immd/immd_db.c b/osaf/services/saf/immsv/immd/immd_db.c
--- a/osaf/services/saf/immsv/immd/immd_db.c
+++ b/osaf/services/saf/immsv/immd/immd_db.c
@@ -296,7 +296,8 @@ uint32_t immd_get_slot_and_subslot_id_fr
 
 	m_NCS_GET_PHYINFO_FROM_NODE_ID(m_NCS_NODE_ID_FROM_MDS_DEST(dest), NULL, &phy_slot, &sub_slot);
 
-	return ((sub_slot * NCS_SUB_SLOT_MAX) + (phy_slot));
+	osafassert(sub_slot < NCS_SUB_SLOT_MAX);
+	return ((uint32_t) sub_slot) + ((uint32_t) phy_slot) * NCS_SUB_SLOT_MAX;
 }
 
 /*****************************************************************************
@@ -313,7 +314,8 @@ uint32_t immd_get_slot_and_subslot_id_fr
 
 	m_NCS_GET_PHYINFO_FROM_NODE_ID(node_id, NULL, &phy_slot, &sub_slot);
 
-	return ((sub_slot * NCS_SUB_SLOT_MAX) + (phy_slot));
+	osafassert(sub_slot < NCS_SUB_SLOT_MAX);
+	return ((uint32_t) sub_slot) + ((uint32_t) phy_slot) * NCS_SUB_SLOT_MAX;
 }
 
 void immd_db_save_fevs(IMMD_CB *cb, IMMSV_FEVS *fevs_msg)
diff --git a/osaf/services/saf/mqsv/mqd/mqd_mds.c b/osaf/services/saf/mqsv/mqd/mqd_mds.c
--- a/osaf/services/saf/mqsv/mqd/mqd_mds.c
+++ b/osaf/services/saf/mqsv/mqd/mqd_mds.c
@@ -567,7 +567,7 @@ static void mqd_mds_svc_evt(MQD_CB *pMqd
 	case NCSMDS_UP:
 		{
 			if (svc->i_svc_id == NCSMDS_SVC_ID_MQND) {
-				uint16_t to_dest_slotid, o_msg_fmt_ver;
+				uint32_t to_dest_slotid, o_msg_fmt_ver;
 				to_dest_slotid = mqsv_get_phy_slot_id(svc->i_dest);
 
 				o_msg_fmt_ver = m_NCS_ENC_MSG_FMT_GET(svc->i_rem_svc_pvt_ver,
diff --git a/osaf/services/saf/smfsv/smfd/smfd_cb.h b/osaf/services/saf/smfsv/smfd/smfd_cb.h
--- a/osaf/services/saf/smfsv/smfd/smfd_cb.h
+++ b/osaf/services/saf/smfsv/smfd/smfd_cb.h
@@ -41,7 +41,6 @@ typedef struct smfd_cb {
 	SaImmOiHandleT campaignOiHandle;            /* IMM Campaign OI handle                           */
 	SaSelectionObjectT campaignSelectionObject; /* Selection Object to wait for campaign IMM events */
 	SaAmfHAStateT ha_state;	                    /* present AMF HA state of the component            */
-	MDS_DEST smfnd_dests[NCS_MAX_SLOTS];        /* destinations for all smfnd   */
 	char *backupCreateCmd;	                    /* Backup create cmd string     */
 	char *bundleCheckCmd;	                    /* Bundle check cmd string      */
 	char *nodeCheckCmd;	                    /* Node check cmd string        */
diff --git a/osaf/tools/safimm/immxml/immxml-nodegen b/osaf/tools/safimm/immxml/immxml-nodegen
--- a/osaf/tools/safimm/immxml/immxml-nodegen
+++ b/osaf/tools/safimm/immxml/immxml-nodegen
@@ -22,7 +22,7 @@ PL_TEMPLATE_FILE=input/imm_PL_template.x
 PL_PATTERN=PL_2_3
 CUR_SC_COUNT=0
 CUR_PL_COUNT=0
-UNDEF_COUNT=999
+UNDEF_COUNT=4093
 SC_COUNT=$UNDEF_COUNT
 PL_COUNT=$UNDEF_COUNT
 RESULT_DIR=./nodegen_result

